src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp

```diff
@@ -0,0 +1,331 @@
+    : center_(center),
+      radius_([&]() noexcept {
+        // The equal_within_roundoff below has an implicit scale of 1,
+        // so the ASSERT may trigger in the case where we really
+        // want an entire domain that is very small.
+        ASSERT(not equal_within_roundoff(radius, 0.0),
+               "Cannot have zero radius");
+        return radius;
+      }()),
+      theta_max_([&]() noexcept {
+        const double cos_theta_max = (z_plane - center_[2]) / radius_;
+        ASSERT(abs(cos_theta_max) < 1.0,
+               "Plane must intersect sphere, and at more than one point");
+        return acos(cos_theta_max);
+      }()) {}
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> Endcap::operator()(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& xbar = source_coords[0];
+  const ReturnType& ybar = source_coords[1];
+  const ReturnType rhobar = sqrt(square(xbar) + square(ybar));
+  const ReturnType sin_factor =
+      radius_ * Endcap_detail::sin_ax_over_x(rhobar, theta_max_);
+  ReturnType z = radius_ * cos(rhobar * theta_max_) + center_[2];
+  ReturnType x = sin_factor * xbar + center_[0];
+  ReturnType y = sin_factor * ybar + center_[1];
+  return std::array<ReturnType, 3>{{std::move(x), std::move(y), std::move(z)}};
+}
+
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame> Endcap::jacobian(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& xbar = source_coords[0];
+  const ReturnType& ybar = source_coords[1];
+  const ReturnType rhobar = sqrt(square(xbar) + square(ybar));
+  const ReturnType sin_factor =
+      radius_ * Endcap_detail::sin_ax_over_x(rhobar, theta_max_);
+  const ReturnType d_sin_factor =
+      radius_ * Endcap_detail::dlogx_sin_ax_over_x(rhobar, theta_max_);
+
+  auto jacobian_matrix =
+      make_with_value<tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>>(
+          dereference_wrapper(source_coords[0]), 0.0);
+
+  // dz/dxbar
+  get<2, 0>(jacobian_matrix) = -sin_factor * theta_max_ * xbar;
+  // dz/dybar
+  get<2, 1>(jacobian_matrix) = -sin_factor * theta_max_ * ybar;
+  // dx/dxbar
+  get<0, 0>(jacobian_matrix) = d_sin_factor * square(xbar) + sin_factor;
+  // dx/dybar
+  get<0, 1>(jacobian_matrix) = d_sin_factor * xbar * ybar;
+  // dy/dxbar
+  get<1, 0>(jacobian_matrix) = get<0, 1>(jacobian_matrix);
+  // dy/dybar
+  get<1, 1>(jacobian_matrix) = d_sin_factor * square(ybar) + sin_factor;
+
+  return jacobian_matrix;
+}
+
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame> Endcap::inv_jacobian(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& xbar = source_coords[0];
+  const ReturnType& ybar = source_coords[1];
+  const ReturnType rhobar = sqrt(square(xbar) + square(ybar));
```

Comment:
I'm not sure if you'll be able to get rid of _all_ the extra allocations, but I would guess at least some of them